Fifo rework

Dolu1990 opened the pull request:
Merge #1144 StreamFifo / StreamFifoLowLatency reimplementation for better perf

Dolu1990 committed:
#1144 Add StreamFifoV2, tested a bit, but not enough

Dolu1990 committed:
add StreamFifoV2 assert

Dolu1990 committed:
#1144 StreamFifoV2 cleaning

Dolu1990 committed:
#1144 StreamFifoV2 now remplace the original StreamFifo (fixed the divergent occupancy calculation)

Dolu1990 committed:
Speedup fifo tester

Dolu1990 committed:
#1144 add allowExtraMsb parameter, cleanup

Dolu1990 committed:
#1144 implement bypass for 1 depth StreamFifo

Dolu1990 committed:
Add Bool.mux(t,f)

Dolu1990 committed:
#1144 Add StreamFifo.forFMax

Dolu1990 committed:
StreamFifo forFMax availabilty has now its own counter

Dolu1990 committed:
Moaaaar fmax for vivado

Dolu1990 committed:
#1144 Implement https://github.com/SpinalHDL/SpinalHDL/issues/1144#issuecomment-1607608182

Dolu1990 committed:
#1144 Fifo formalCheckRam now preserve names

Dolu1990 committed:
#1144 Fix broken code

fpgahouse committed:
Stream.scala: Document formalAssumesOrder() code.

fpgahouse committed:
FormalFifoTester.scala: Comment for empty/fullTracker asserts

fpgahouse committed:
Stream.scala: Fix formalCheckm2sPipe() for !withAsyncRead

fpgahouse committed:
FormalFifoTester.scala: Prepare for other StreamFifo() configs.

fpgahouse committed:
Rename formalCheckm2sPipe to  formalCheckOutputStage.

Dolu1990 committed:
#1144 StreamFifo.useVec added

Dolu1990 committed:
#1144 Good bye StreamFifoLowLatency aswell <3

Dolu1990 committed:
#1144 better checks

Dolu1990 committed:
#1144 remove StreamFifo.occupancyFromRamOnly

fpgahouse committed:
Merge remote-tracking branch 'upstream/fifo-rework' into fifo-rework

fpgahouse committed:
FormalFifoTester.scala: Iterate through configurations.

fpgahouse committed:
FormalFifoTester.scala: Remove occupancyFromRamOnly.

Dolu1990 committed:
Merge pull request #1146 from likewise/fifo-rework

Improve formal verification on fifo-rework

fpgahouse committed:
FormalFifoTester.scala: More coverage, document TODO.

fpgahouse committed:
FormalFifoTester.scala: Back-2-back test enabled only for depth>=4 for sync.

andreasWallner commented on lib/src/main/scala/spinal/lib/Stream.scala:
```diff
@@ -1085,101 +1090,277 @@ object StreamFifo{
 
-class StreamFifo[T <: Data](val dataType: HardType[T], val depth: Int) extends Component {
+/**
+  * Improved StreamFifo over the original implementation
```
not sure we need to refer to the original one that does not exist anymore after this merge?

likewise replied:
Maybe mention: Fully redesigned in release 1.8.2 allowing improved timing closure.

Dolu1990 replied:
yes, doing it.

andreasWallner commented on lib/src/main/scala/spinal/lib/Stream.scala:
```diff
@@ -1085,101 +1090,277 @@ object StreamFifo{
   require(depth >= 0)
-  val io = new Bundle {
+  // bypass not supported for sync read case
+  if(!withAsyncRead) require(!withBypass)
```
should we add a note like "use of withAsyncRead=false requires withBypass=false as well"?

likewise replied:
Yes. Same for `useVec`, as it requires` withAsyncRead`.
We should add this to the code:
```scala
if (useVec) require (withAsyncRead)
```

Dolu1990 replied:
Yes, on it.

andreasWallner commented on tester/src/test/scala/spinal/tester/scalatest/FormalFifoTester.scala:
```diff
@@ -44,23 +61,58 @@ class FormalFifoTester extends SpinalFormalFunSuite {
+        // cover the case that FIFO first goes to full, then to empty
+        dut.formalFullToEmpty()
+
+        // TODO for .withProve() induction prove to pass, relationship between push/pop distance and some fill rate
```
are the TODOs in here (and the one below on line 124) still a thing?

likewise replied:
Yes, the TODO's are up-to-date with the code.

I have newer formal code with update TODO's in my `fifo-rework` branch fork, matching better formal coverage.

andreasWallner commented on tester/src/test/scala/spinal/tester/scalatest/SpinalSimStreamFifoTester.scala:
```diff
@@ -31,31 +40,68 @@ class SpinalSimStreamFifoTester extends SpinalSimFunSuite {
       dut.io.push.valid #= false
       dut.clockDomain.onSamplings {
         if (dut.io.push.valid.toBoolean && dut.io.push.ready.toBoolean) {
           queueModel.enqueue(dut.io.push.payload.toLong)
```
since we update the testcase, should we use `ScoreboardInOrder` here?

Dolu1990 replied:
Baaaa, things are working already, i think it isn't necessary.

andreasWallner reviewed (commented):
Just a few general comments on minor stuff...

One open point in my mind: should some of the parameters also be available from the `Stream.queue` function?

fpgahouse committed:
More readable code: Use brackets in (1 << (a - b)), was (1 << a - b)

fpgahouse committed:
StreamFifo configurations: Better formal verification coverage.

forFMax fully passes all formal tests.
!forFMax passes BMC and Cover formal tests, the Prove tests requires more formal test code.

This passes with the above restrictions:

  for (depth <- List(3, 4, 7);
    withAsyncRead <- List(false, true);
    withBypass <- List(false, true);
    forFMax <- List(false, true);
    allowExtraMsb <- List(false, true);

fpgahouse committed:
StreamFifo: formalCheckRam now supports useVec also.

fpgahouse committed:
FormalFifoTester.scala: Use dataWidth, and allow useVec testing.

Dolu1990 committed:
Merge pull request #1150 from likewise/fifo-rework

Fifo rework

Dolu1990 commented:
> One open point in my mind: should some of the parameters also be available from the `Stream.queue` function?

Yes, i would say, latency and forFMax tunning

Dolu1990 committed:
Better doc / asserts

Dolu1990 committed:
Add more StreamFIfo API
